### PR TITLE
Shipping Labels: Save changes to payment methods

### LINKED ISSUE #4086: Shipping Labels: Payment Method
Add ability to choose a payment method when creating a shipping label, this includes:

- [x] On the `Create Shipping Label` screen, display the selected payment method ("Credit card ending in ####"). By default this is identified by `selectedPaymentMethodID` in `ShippingLabelAccountSettings`, unless a different method is selected in this flow. (Option to add a new card is not part of M3.) https://github.com/woocommerce/woocommerce-ios/pull/4256
- [x] List the currently added payment methods, and the selected one if there is any. https://github.com/woocommerce/woocommerce-ios/pull/4256
- [x] Allow selecting a payment method from the list.
- [ ] Remotely update the shipping label account settings with changes to the payment method or email receipts setting. (Updating the payment method is required to use the newly selected payment method for the shipping label purchase.)

Feel free to split this task into sub-tasks.

| Payment Method - start of the flow            |  Credit card selection |
:-------------------------:|:-------------------------:
Schermata 2021-04-30 alle 12 40 48 | Schermata 2021-04-30 alle 12 42 10

### PR DESCRIPTION
**Pull request by rachelmcr**
Closes: #4086 

## Description

This PR makes the Done button operational on the Payment Method screen, to save changes to the payment method settings. After tapping Done, the changes are updated remotely (using the endpoint added in #4310) so they can be used when the shipping label is purchased.

## Changes

* UI:
   * Enables the Done button on the Payment Method screen when there are changes to the account settings (selected payment method or the email receipts setting).
   * Fires the API request to update the account settings remotely when the Done button is tapped, and displays an activity indicator while the update is in progress.
   * Displays an error notice if the remote update fails.
   * Passes the updates settings back to the `ShippingLabelFormViewController` so the change can be handled there (displaying the updated credit card details).
* Networking:
   * Fixes the remote endpoint to use the correct response mapper (a success response wrapped in a `data` key).
* Testing:
   * Makes `ShippingLabelPaymentMethod` conform to `GeneratedCopiable` and adds copiable methods for use in unit tests.
   * Updates existing unit tests and adds new unit tests for the relevant view models (`ShippingLabelFormViewModel` and `ShippingLabelPaymentMethodsViewModel`).

## Testing

**Setup**

Make sure you have at least two payment methods set up in your store's WP Admin under WooCommerce > Settings > Shipping > WooCommerce Shipping (/wp-admin/admin.php?page=wc-settings&tab=shipping&section=woocommerce-services-settings). The ability to add new payment methods in the app will be implemented in the next milestone.

**Case: Successful update**

1. Launch the app in debug mode (because of the feature flag).
2. Open an order detail that is eligible for the shipping label, where the customer paid a flat fee for the shipping.
3. Press the button "Create Shipping Label".
4. Press the button "Continue" in Create Shipping Label form, under the **Ship From** cell. Confirm the address.
5. Press the button "Continue" in Create Shipping Label form, under the **Ship To** cell. Confirm the address.
6. Press the button "Continue" in Create Shipping Label form, under the **Package Details** cell.
7. You should see the screen that shows the Package Details screen and all the items that will be shipped.
8. Press the "Package selected" row.
9. Select a package, then press "Done".
10. Add a Total package weight, press "Done".
11. The "Shipping Carrier and Rates" cell should be enabled now. Press it.
12. Select a carrier, then press "Done".
13. The "Payment Methods" cell should be enabled now, with your selected payment method details displayed. Press it.
14. On the Payment Method screen, confirm the "Done" button is not enabled.
15. Select a different payment method (and optionally change the email receipts setting), then press "Done".
16. Confirm you see an activity indicator (spinner) in place of the "Done" button while the update is happening.
17. Confirm the Payment Method cell now displays the digits for the new payment method you selected.
18. In WP Admin, go to WooCommerce > Settings > Shipping > WooCommerce Shipping and confirm that your changes were saved remotely.

**Case: Failed update**

1. Repeat the steps 1-14 above (or just reselect the Payment Method cell after finishing that test case).
2. On the Payment Method screen, select a different payment method and/or change the email receipts setting.
3. Disable your network connection, then press "Done".
4. Confirm an error notice appears informing you that the update failed, and you remain on the Payment Method screen.

## Submitter Checklist

Update release notes:
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipping Labels/Create Shipping Label Form/Payment Methods/ShippingLabelPaymentMethods.swift [hunk 1]
```diff
@@ -43,9 +51,19 @@ struct ShippingLabelPaymentMethods: View {
...
-            Text(Localization.doneButton)
-        }))
+        .navigationBarItems(trailing: Button(action: {
+            viewModel.updateShippingLabelAccountSettings {newSettings in
+                onCompletion(newSettings)
+                presentation.wrappedValue.dismiss()
+            }
+        }, label: {
+            if viewModel.isUpdating {
+                ProgressView()
```
Comment (pmusolino) [inline WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipping Labels/Create Shipping Label Form/Payment Methods/ShippingLabelPaymentMethods.swift:61]: I just found that the spinner indicator in the navigation bar remains tappable while loading. I think we should disable it while we are in the loading process. Feel free to fix it in this PR or in a new one. 👍

File: WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipping Labels/Create Shipping Label Form/Payment Methods/ShippingLabelPaymentMethods.swift [hunk 2]
```diff
@@ -43,9 +51,19 @@ struct ShippingLabelPaymentMethods: View {
         }
         .background(Color(.listBackground))
         .navigationBarTitle(Localization.navigationBarTitle)
-        .navigationBarItems(trailing: Button(action: {}, label: {
-            Text(Localization.doneButton)
-        }))
+        .navigationBarItems(trailing: Button(action: {
+            viewModel.updateShippingLabelAccountSettings {newSettings in
```
Comment (pmusolino) [inline WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipping Labels/Create Shipping Label Form/Payment Methods/ShippingLabelPaymentMethods.swift]:
```suggestion
            viewModel.updateShippingLabelAccountSettings { newSettings in
```

File: WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipping Labels/Create Shipping Label Form/ShippingLabelFormViewController.swift [hunk 3]
```diff
@@ -313,9 +313,16 @@ private extension ShippingLabelFormViewController {
...
+            return
+        }
+
+        let vm = ShippingLabelPaymentMethodsViewModel(accountSettings: accountSettings)
+        let paymentMethod = ShippingLabelPaymentMethods(viewModel: vm) { [weak self] newSettings in
+            guard let self = self else { return }
+            self.viewModel.handlePaymentMethodValueChanges(settings: newSettings, editable: true)
+        }
+
         let hostingVC = UIHostingController(rootView: paymentMethod)
```
Comment (pmusolino) [inline WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipping Labels/Create Shipping Label Form/ShippingLabelFormViewController.swift]:
```suggestion
```

File: WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipping Labels/Create Shipping Label Form/ShippingLabelFormViewModel.swift [hunk 4]
```diff
@@ -147,11 +149,12 @@ final class ShippingLabelFormViewModel {
         updateRowState(type: .shippingCarrierAndRates, dataState: .validated, displayMode: .editable)
     }
 
-    func handlePaymentMethodValueChanges(selectedPaymentMethodID: Int64, editable: Bool) {
-        self.selectedPaymentMethodID = selectedPaymentMethodID
+    func handlePaymentMethodValueChanges(settings: ShippingLabelAccountSettings, editable: Bool) {
+        self.shippingLabelAccountSettings = settings
```
Comment (pmusolino) [inline WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipping Labels/Create Shipping Label Form/ShippingLabelFormViewModel.swift]: nit: `self` is not needed

File: WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipping Labels/Create Shipping Label Form/Payment Methods/ShippingLabelPaymentMethodsViewModel.swift [hunk 5]
```diff
@@ -6,45 +6,92 @@ import protocol Storage.StorageManagerType
...
+
+    /// Updates remote shipping label account settings
+    ///
+    func updateShippingLabelAccountSettings(onCompletion: @escaping ((ShippingLabelAccountSettings) -> Void)) {
+        isUpdating = true
+        let newSettings = accountSettings.copy(selectedPaymentMethodID: selectedPaymentMethodID,
+                                        isEmailReceiptsEnabled: isEmailReceiptsEnabled)
+
+        let action = ShippingLabelAction.updateShippingLabelAccountSettings(siteID: accountSettings.siteID, settings: newSettings) { result in
+            self.isUpdating = false
```
Comment (pmusolino) [inline WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipping Labels/Create Shipping Label Form/Payment Methods/ShippingLabelPaymentMethodsViewModel.swift]:
```suggestion
        let action = ShippingLabelAction.updateShippingLabelAccountSettings(siteID: accountSettings.siteID, settings: newSettings) { [weak self] result in
            self?.isUpdating = false
```

File: WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipping Labels/Create Shipping Label Form/Payment Methods/ShippingLabelPaymentMethodsViewModel.swift [hunk 6]
```diff
@@ -6,45 +6,92 @@ import protocol Storage.StorageManagerType
...
+                DDLogError("⛔️ Error updating shipping label account settings")
+            }
+        }
+        ServiceLocator.stores.dispatch(action)
+    }
+}
+
+// MARK: - Localization
+//
+extension ShippingLabelPaymentMethodsViewModel {
```
Comment (pmusolino) [inline WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipping Labels/Create Shipping Label Form/Payment Methods/ShippingLabelPaymentMethodsViewModel.swift]:
```suggestion
private extension ShippingLabelPaymentMethodsViewModel {
```

**Review by pmusolino — APPROVED** [review]
Looks good to me :shipit: 

Code ✅ just some minor nits
Testing ✅

File: WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipping Labels/Create Shipping Label Form/Payment Methods/ShippingLabelPaymentMethods.swift (hunk 1 above)
Comment (rachelmcr) [reply WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipping Labels/Create Shipping Label Form/Payment Methods/ShippingLabelPaymentMethods.swift:61]: Good catch, thanks! Fixed that in 65e1cbc.

**Comment by rachelmcr** [comment]
Thanks for the review, @pmusolino! I really appreciate your attention to detail. :)

**Pull request merged**